src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -1109,6 +1110,146 @@ int RadosStore::remove_topics(const std::string& tenant, RGWObjVersionTracker* o
+                                /*exclusive*/ false);
+  std::unique_ptr<RGWSI_MetaBackend::Context> ctx(
+      svc()->topic->svc.meta_be->alloc_ctx());
+  ctx->init(svc()->topic->get_be_handler());
+  return svc()->topic->svc.meta_be->put(
+      ctx.get(), get_topic_key(topic.name, topic.user.tenant), params,
+      objv_tracker, y, dpp);
+}
+
+int RadosStore::remove_topic_v2(const std::string& topic_name,
+                                const std::string& tenant,
+                                RGWObjVersionTracker* objv_tracker,
+                                optional_yield y,
+                                const DoutPrefixProvider* dpp) {
+  RGWSI_MBSObj_RemoveParams params;
+  std::unique_ptr<RGWSI_MetaBackend::Context> ctx(
+      svc()->topic->svc.meta_be->alloc_ctx());
+  ctx->init(svc()->topic->get_be_handler());
+  return svc()->topic->svc.meta_be->remove(ctx.get(),
+                                           get_topic_key(topic_name, tenant),
+                                           params, objv_tracker, y, dpp);
+}
+
+int RadosStore::update_bucket_topic_mapping(const rgw_pubsub_topic& topic,
+                                            const std::string& bucket_key,
+                                            bool add_mapping,
+                                            optional_yield y,
+                                            const DoutPrefixProvider* dpp) {
+  bufferlist empty_bl;
+  librados::ObjectWriteOperation op;
+  int ret = 0;
+  if (add_mapping) {
+    std::map<std::string, bufferlist> mapping{{bucket_key, empty_bl}};
+    op.omap_set(mapping);
+  } else {
+    std::set<std::string> to_rm{{bucket_key}};
+    op.omap_rm_keys(to_rm);
+  }
+  ret = rgw_rados_operate(dpp, rados->get_notif_pool_ctx(),
+                          get_bucket_topic_mapping_oid(topic), &op, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, 1) << "ERROR: failed to " << (add_mapping ? "add" : "remove")
+                      << " topic bucket mapping for bucket: " << bucket_key
+                      << " and topic: " << topic.name << " with ret:" << ret << dendl;
+    return ret;
+  }
+  ldpp_dout(dpp, 20) << "Successfully " << (add_mapping ? "added" : "removed")
+                     << " topic bucket mapping for bucket: " << bucket_key
+                     << " and topic: " << topic.name << dendl;
+  return ret;
+}
+
+int RadosStore::get_bucket_topic_mapping(const rgw_pubsub_topic& topic,
+                                         std::set<std::string>& bucket_keys,
+                                         optional_yield y,
+                                         const DoutPrefixProvider* dpp) {
+  constexpr auto max_chunk = 1024U;
+  std::string start_after;
+  bool more = true;
+  int rval;
+  while (more) {
+    librados::ObjectReadOperation op;
+    std::set<std::string> curr_keys;
+    op.omap_get_keys2(start_after, max_chunk, &curr_keys, &more, &rval);
+    const auto ret =
```

Comment:
it doesn't look like the `start_after` marker gets updated after each response, so this loop wouldn't terminate if there are more than 1024 entries:
```diff
     }
+    if (!curr_keys.empty()) {
+      start_after = *curr_keys.rbegin();
+    }
     bucket_keys.merge(curr_keys);
  }
```
(cc @yuvalif, it looks like `RadosLuaManager::list_packages()` and `rgw::notify::Manager::read_queue_list()` have copy/pasted this same pattern)

Reply:
done.